sshd-core/src/main/java/org/apache/sshd/server/forward/TcpipServerChannel.java

```diff
@@ -217,9 +225,23 @@ public void messageReceived(IoSession session, Readable message) throws Exceptio
                         log.debug("doInit({}) Ignoring write to channel in CLOSING state", TcpipServerChannel.this);
                     }
                 } else {
-                    Buffer buffer = new ByteArrayBuffer(message.available(), false);
+                    int length = message.available();
+                    Buffer buffer = new ByteArrayBuffer(length, false);
                     buffer.putBuffer(message);
-                    out.writePacket(buffer);
+                    long total = inFlightDataSize.addAndGet(length);
+                    if (total > maxBufferSize) {
+                        session.suspendRead();
+                    }
+                    IoWriteFuture ioWriteFuture = out.writePacket(buffer);
+                    ioWriteFuture.addListener(new SshFutureListener<IoWriteFuture>() {
+                        @Override
+                        public void operationComplete(IoWriteFuture future) {
+                            long total = inFlightDataSize.addAndGet(-length);
+                            if (total <= maxBufferSize) {
```

Comment:
Maybe this threshold can be configured?
when maxBufferSize = 1MB,  I think ```if (total <= 500KB)``` might be better for performance? to avoid frequent suspensions and resumptions

inFlightDataSize > 1 MB,  suspendRead
inFlightDataSize < 0.5 MB, resumedRead

Reply:
Right, that makes sense, I'll split the property.